python/src/uagents/agent.py

```diff
@@ -961,7 +963,52 @@ async def _process_message_queue(self):
             )
             if handler is None:
                 if not is_user_address(sender):
-                    handler = self._signed_message_handlers.get(schema_digest)
+                    is_valid = True  # always valid unless considered invalid as part of a dialogue
+                    for protocol in context.protocols.values():
+                        if hasattr(protocol, "rules") and protocol.is_included(
+                            schema_digest
+                        ):
+                            state = protocol.get_current_state(session)
+                            context.logger.debug(
+                                "current state: "
+                                f"{(protocol.models[state].__name__ if state else 'n/a')}"
+                            )
+                            is_valid = protocol.is_valid_message(session, schema_digest)
+                            context.logger.debug(
+                                f"message {self._models[schema_digest].__name__} "
+                                f"allowed: {is_valid}"
+                            )
+
+                            if not is_valid:
+                                context.reset_session()
+                                await _send_error_message(
+                                    context,
+                                    sender,
+                                    ErrorMessage(
+                                        error=f"Unexpected message in dialogue: {message}"
+                                    ),
+                                )
+                                break
+
+                            if protocol.is_starter(schema_digest):
+                                self._ctx.logger.debug("dialogue started")
+                            elif protocol.is_ender(schema_digest):
+                                self._ctx.logger.debug(
+                                    "dialogue ended, cleaning up session"
+                                )
+                            else:
+                                self._ctx.logger.debug("dialogue picked up")
+
+                            context.dialogue = protocol.get_session(session)
+                            protocol.add_message(
+                                session_id=session,
+                                message_type=self._models[schema_digest].__name__,
+                                sender=sender,
+                                receiver=self.address,
+                                content=message,
+                            )
```

Comment:
Can we add this logic to the `Dialogue` class instead of putting it here in `Agent`?

Reply:
I think this is one of the comments I made once before. It seems I was unconvinced. :smile:

Reply:
This logic here and in context that is actually only relevant to dialogues was the main reason to consider a greater refactor of the `Protocol` <> `Agent` <> `Context` interaction / dependencies. Of course we could reduce the footprint here by moving some code to the `Dialogue` class, but originally we left this here to actually have this discussion

Reply:
Exactly this has to do with the bigger refactor. Ideally we would try to put this code snippet in the `Dialogue` class but this is one of two instances where we need to directly tap into the uagents core mechanics so I don't see a way around that at the moment.